src/io/flutter/jxbrowser/JxBrowserManager.java

```diff
@@ -0,0 +1,209 @@
+import com.intellij.util.download.DownloadableFileService;
+import com.intellij.util.download.FileDownloader;
+import com.intellij.util.lang.UrlClassLoader;
+import io.flutter.utils.FileUtils;
+import io.flutter.utils.JxBrowserUtils;
+import org.jetbrains.annotations.NotNull;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicReference;
+
+enum JxBrowserStatus {
+  NOT_INSTALLED,
+  INSTALLATION_IN_PROGRESS,
+  INSTALLED,
+  INSTALLATION_FAILED,
+}
+
+// JxBrowser provides Chromium to display web pages within IntelliJ. This class manages downloading the required files and adding them to
+// the class path.
+public class JxBrowserManager {
+  private static JxBrowserManager manager;
+  private static final String DOWNLOAD_PATH = FileUtils.platformPath();
+  private static final AtomicReference<JxBrowserStatus> status = new AtomicReference<>(JxBrowserStatus.NOT_INSTALLED);
+  private static final Logger LOG = Logger.getInstance(JxBrowserManager.class);
+  // We will be gating JxBrowser features until all of the features are landed.
+  // To test JxBrowser, set this to true and also add license key to VM options (-Djxbrowser.license.key=<key>).
+  public static final boolean ENABLE_JX_BROWSER = false;
+
+  private JxBrowserManager() {}
+
+  public static JxBrowserManager get() {
+    if (manager == null) {
+      return new JxBrowserManager();
+    }
+    return manager;
+  }
+
+  public boolean isInstalled() {
+    return status.get().equals(JxBrowserStatus.INSTALLED);
+  }
+
+  public void setUp(Project project) {
+    if (!status.compareAndSet(JxBrowserStatus.NOT_INSTALLED, JxBrowserStatus.INSTALLATION_IN_PROGRESS)) {
+      // This check ensures that an IDE only downloads and installs JxBrowser once, even if multiple projects are open.
+      // If already in progress, let calling point wait until success or failure (it may make sense to call setUp but proceed).
+      // If already succeeded or failed, no need to continue.
+      return;
+    }
+
+    LOG.info(project.getName() + ": Installing JxBrowser");
+
+    final File directory = new File(DOWNLOAD_PATH);
+    if (!directory.exists()) {
+      if (!directory.mkdirs()) {
+        LOG.info(project.getName() + ": Unable to create directory for JxBrowser files");
+        status.set(JxBrowserStatus.INSTALLATION_FAILED);
+        return;
+      }
+    }
+
+    // Check for file or loading file
+    final String[] fileNames = {JxBrowserUtils.getPlatformFileName(), JxBrowserUtils.getApiFileName(), JxBrowserUtils.getSwingFileName()};
+    boolean allDownloaded = true;
+    final List<File> files = new ArrayList<>();
+    for (String fileName : fileNames) {
+      final File file = new File(DOWNLOAD_PATH + File.separator + fileName);
+      files.add(file);
+      if (!file.exists()) {
+        allDownloaded = false;
+        break;
```

Comment:
this break seems wrong. The list of `files` will sometimes be missing a file.

Reply:
Wow good catch. Could be helpful to add unit tests in a next change.